doc/shared/sunheuristics/SUNHeuristics_Description.rst

```diff
@@ -0,0 +1,647 @@
+                                         realtype hnew, realtype *hconstr);
+      int              (*boundfirststep)(SUNHeuristics H, realtype h0,
+                                         realtype *h0constr);
+      int              (*reset)(SUNHeuristics H);
+      int              (*update)(SUNHeuristics H);
+      int              (*setdefaults)(SUNHeuristics H);
+      int              (*write)(SUNHeuristics H, FILE* fptr);
+      int              (*setmaxstep)(SUNHeuristics H, realtype hmax);
+      int              (*setminstep)(SUNHeuristics H, realtype hmin);
+      int              (*setexpstabfn)(SUNHeuristics H, SUNExpStabFn EStab,
+                                       void* estab_data);
+      int              (*setcflfraction)(SUNHeuristics H, realtype cfl_frac);
+      int              (*setsafetyfactor)(SUNHeuristics H, realtype safety);
+      int              (*setmaxgrowth)(SUNHeuristics H, realtype mx_growth);
+      int              (*setminreduction)(SUNHeuristics H, realtype eta_min);
+      int              (*setfixedstepbounds)(SUNHeuristics H, realtype lb, realtype ub);
+      int              (*setmaxfirstgrowth)(SUNHeuristics H, realtype etamx1);
+      int              (*setmaxefailgrowth)(SUNHeuristics H, realtype etamxf);
+      int              (*setsmallnumefails)(SUNHeuristics H, int small_nef);
+      int              (*setmaxcfailgrowth)(SUNHeuristics H, realtype etacf);
+      int              (*getnumexpsteps)(SUNHeuristics H, long int* expsteps);
+      int              (*getnumaccsteps)(SUNHeuristics H, long int* accsteps);
+      int              (*space)(SUNHeuristics H, long int *lenrw, long int *leniw);
+    };
+
+
+SUNHeuristics Type Definitions
+------------------------------
+
+The time integrators in SUNDIALS leverage a myriad of heuristics to achieve
+accurate and efficient computations; however, most follow a rather standard
+pattern, or request no heuristic control whatsoever.  Therefore, each
+SUNHeuristics implementation should indicate its type, informing the integrators
+of whether they perform heuristic control or not:
+
+.. _SUNHeuristics.Description.heuristicsIDs:
+.. table:: Identifiers associated with SUNHeuristics implementations
+   :align: center
+
+   ========================  =====================================================
+   SUNHeuristics ID          Intended form of heuristic control
+   ========================  =====================================================
+   SUNDIALS_HEURISTICS_STD   Performs "standard" heuristic stepsize constraints.
+   SUNDIALS_HEURISTICS_NULL  Performs no heuristic constraints.
+   ========================  =====================================================
+
+
+Additionally, if the heuristics object can be used to limit stepsizes based on
+explicit stability, then we define the following user-supplied function type:
+
+.. c:type:: int (*SUNExpStabFn)(realtype *hstab, void *user_data)
```

Comment:
Stability is orthogonal to accuracy, and so I do not consider it to be part of the error controller, but instead an heuristic constraint on the allowable time step.

Furthermore, as currently organized, it is perfectly feasible to combine both this stability restriction to the heuristics, while also using temporal error control.  Then, a problem can transition between accuracy-limited and stability-limited steps seamlessly.  We cannot, however, currently combine two error controllers within the same calculation, so if the explicit stability were a separate `SUNAdaptController`, it could NOT be combined with a true temporal error controller.

Reply:
I guess I'm thinking about controller objects more generally as any method for adjusting the step size/order instead of only an error-based controller. So a CFL controller would just be another "type" of controller (e.g., `SUN_Controller_CFL`) that gets attached with a corresponding set function (e.g., `ARKStepSetCFLController`).

This could similarly be applied to the heuristic controller to break it up into smaller controllers with a more specific purpose (something like `SUN_Controller_H_Bounds`, `SUN_Controller_H_Growth`, etc.) but this would require several implementations and I'm not sure that's necessary in order to turn off the heuristics that are problematic for enabling adaptivity in SPRK step.